src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -243,6 +245,25 @@ public override List<TSource> ToList()
                 return list;
             }
 
+            public override HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                HashSet<TSource> hashSet = new HashSet<TSource>(comparer);
+
+                if (!_appending)
+                {
+                    hashSet.Add(_item);
+                }
+
+                hashSet.UnionWith(_source);
+
+                if (_appending)
```

Comment:
Given the nature of hash sets we don't care if it's at the beginning or the end, so no need to check on `_appending` to decide when to add, just `HashSet<TSource> hashSet = new HashSet<TSource>(comparer); hashSet.Add(_item)`.